### PR TITLE
Scripts for feature "Mobile Projection phase 2"

### PR DESCRIPTION
**Pull request by dboltovskyi**
ATF Test Scripts to check #[2129](https://github.com/smartdevicelink/sdl_core/issues/2129)

This PR is **ready** for review.

### Summary
Scripts for feature "Mobile Projection phase 2".

### ATF version
[Develop branch](https://github.com/smartdevicelink/sdl_atf/tree/develop)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by yang1070** [comment]
@theresalech Ford has approved this PR

**Comment by JackLivio** [comment]
@dboltovskyi Hello it looks like the test cases are missing their normal descriptions/comments?

**Comment by JackLivio** [comment]
Also these test cases take a very long time to complete. We should look at a way to optimize these test scripts or decide if all of these test cases are necessary for this new feature.

**Comment by dboltovskyi** [comment]
@JackLivio Descriptions were missed due to lack of time and will be provided in upcoming additional commit.

**Comment by dboltovskyi** [comment]
@JackLivio Regarding you point on huge number of test cases.
Right, proposal describes the logic just only for:
  - new ```videoStreamingState``` parameter (5 cases)
  - existing ```audioStreamingState``` parameter (11 cases)

But it looks simple than it is. Let me explain the reasons:
1. In cases there is no description of exact values for ```appHMIType``` and ```isMediaApplication``` parameters of mobile application. These parameters are important to identify the A/V states. E.g.: NAVIGATION app could be AUDIBLE and STREAMABLE regardless the value of ```isMediaApplication``` flag. But PROJECTION app takes the value of this parameter into account, so this app could be 2 types:
AUDIBLE-STREAMABLE or NOT_AUDIBLE/STREAMABLE.
2. In cases there is a high level description of app's current state (e.g. use video/audio, does not use video/audio). But it is also important to know current ```hmiLevel``` (e.g. FULL or LIMITED) 
3. Many cases includes various app types (e.g.: 4.2: NAVIGATION or COMMUNICATION, 4.4:Non-NAVIGATION, NON-COMMUNICATION etc.)
4. Some cases includes many events that needs to be checked (e.g. 5.1 PHONE_CALL / EMERGENCY_EVENT / AUDIO_SOURCE / EMBEDDED_NAVI etc.)
5.  Now ```hmiLevel``` depends on current values of ```audioStreamingState``` and ```videoStreamingState``` parameters

There following variables that needs to be taken into account:
  - app's configuration (```appHMIType``` and ```isMediaApplication```)
  - app's initial state
  - number of apps
  - events occurs
  - etc.

All these lead to big amount of combinations.

State controller functionality is one of the most complicated parts of SDL and by implementing this huge number of test cases we tried to cover as much as possible (including not only new logic, but also existing one)

As part of functional testing of this new feature I suggest to run all of the cases.
But for regression I propose to create separate test set with reduced number of them.

File: test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/MobileProjection/Phase2/common')
+local runner = require('user_modules/script_runner')
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local Variables ]]
+local testCases = {
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua]: @JackLivio Could you provide SDL and ATF logs please

File: test_scripts/MobileProjection/Phase2/014_audio_010-068_video_007-030_states_LIMITED.lua [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/MobileProjection/Phase2/common')
+local runner = require('user_modules/script_runner')
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+
+--[[ Local Variables ]]
+local testCases = {
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/MobileProjection/Phase2/014_audio_010-068_video_007-030_states_LIMITED.lua]: @JackLivio Could you provide SDL and ATF logs please

**Comment by JackLivio** [comment]
@dboltovskyi Thank for for the response and information. I understand the need for all of the combinations of tests. I also like the idea of having a smaller alternate test set available for regression.

File: test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (JackLivio) [reply test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua]: Logs are too large to post for this test. Here is the terminal output of failed test cases
```
--- TC[066]: App1[hmiType:DEFAULT, isMedia:true], App2[hmiType:PROJECTION, isMedia:true] ------------
[16:55:19,900] Clean_environment                                                                     [SUCCESS] (2 ms)
SDL pid 26717
tcp        0      0 127.0.0.1:8087          0.0.0.0:*               LISTEN     
HMI initialized
HMI is ready
Mobile connected
[16:55:21,042] Start_SDL__HMI__connect_Mobile__start_Session                                         [FAIL] (19172 ms)
 HMI initialization: Timeout: HMI initialization: Timeout expired
[16:55:40,699] Set_App_1_Config                                                                      [SUCCESS] (7 ms)
[16:55:41,172] Set_App_2_Config                                                                      [SUCCESS] (5 ms)
[16:55:41,291] Register_App_1                                                                        [SUCCESS] (2908 ms)
[16:55:44,649] Register_App_2                                                                        [SUCCESS] (3471 ms)
[16:55:48,618] Activate_App_1                                                                        [SUCCESS] (2223 ms)
[16:55:51,297] Activate_App_2__audioStates__app1_NOT_AUDIBLE__app2_AUDIBLE                           [SUCCESS] (21 ms)
Mobile session 1 deleted
Mobile session 2 deleted
[16:55:51,671] Clean_sessions                                                                        [SUCCESS] (2026 ms)
[16:55:54,899] Stop_SDL                                                                              [SUCCESS] (2 ms)
--- TC[067]: App1[hmiType:DEFAULT, isMedia:true], App2[hmiType:MEDIA, isMedia:true] -----------------
[16:55:58,047] Clean_environment                                                                     [SUCCESS] (3 ms)
SDL pid 27135
tcp        0      0 127.0.0.1:8087          0.0.0.0:*               LISTEN     
HMI initialized
HMI is ready
Mobile connected
[16:55:59,198] Start_SDL__HMI__connect_Mobile__start_Session                                         [FAIL] (19213 ms)
 HMI initialization: Timeout: HMI initialization: Timeout expired
[16:56:18,927] Set_App_1_Config                                                                      [SUCCESS] (5 ms)
[16:56:19,421] Set_App_2_Config                                                                      [SUCCESS] (7 ms)
[16:56:19,555] Register_App_1                                                                        [SUCCESS] (2953 ms)
[16:56:22,990] Register_App_2                                                                        [SUCCESS] (3464 ms)
[16:56:26,930] Activate_App_1                                                                        [SUCCESS] (2024 ms)
[16:56:29,437] Activate_App_2__audioStates__app1_NOT_AUDIBLE__app2_AUDIBLE                           [SUCCESS] (24 ms)
Mobile session 1 deleted
Mobile session 2 deleted
[16:56:29,810] Clean_sessions                                                                        [SUCCESS] (2020 ms)
[16:56:32,980] Stop_SDL                                                                              [SUCCESS] (2 ms)
--- TC[068]: App1[hmiType:DEFAULT, isMedia:true], App2[hmiType:DEFAULT, isMedia:true] ---------------
[16:56:36,118] Clean_environment                                                                     [SUCCESS] (2 ms)
SDL pid 27546
tcp        0      0 127.0.0.1:8087          0.0.0.0:*               LISTEN     
[16:56:37,270] Start_SDL__HMI__connect_Mobile__start_Session                                         [FAIL] (11220 ms)
 HMI initialization: Timeout: HMI initialization: Timeout expired
[16:56:49,003] Set_App_1_Config                                                                      [SUCCESS] (5 ms)
[16:56:49,129] Set_App_2_Config                                                                      [SUCCESS] (262 ms)
[16:56:49,509] Register_App_1                                                                        [FAIL] (10692 ms)
 StartService ACK: Timeout: StartService ACK: Timeout expired
[16:57:00,786] Register_App_2                                                                        [FAIL] (10624 ms)
 StartService ACK: Timeout: StartService ACK: Timeout expired
[16:57:11,989] Activate_App_1                                                                        [FAIL] (10616 ms)
 OnHMIStatus notification: Timeout: OnHMIStatus notification: Timeout expired
[16:57:23,192] Activate_App_2__audioStates__app1_NOT_AUDIBLE__app2_AUDIBLE                           [FAIL] (10736 ms)
 OnHMIStatus notification: Timeout: OnHMIStatus notification: Timeout expired

```

[the remark this answers is not in the document]
Comment (JackLivio) [reply test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua]: If I only run tests 66-68, then they pass. I wonder if there is some instability with ATF that causes failures when a large test set it run?

I have not been able to complete all tests successfully on multiple attempts.

[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/MobileProjection/Phase2/002_audioStreamingState_010-068.lua]: @JackLivio In logs it was found out that there is a performance degradation of execution of  `Start_SDL__HMI__connect_Mobile__start_Session` step.
At the start duration of this step is 1-2 sec. But at the end - 18-20 sec.
That is the reason of ATF `Timeout expired` message.

Usually execution of this step takes no more than 5 sec.
So these issue could be treated as environment related.

File: test_scripts/MobileProjection/Phase2/014_audio_010-068_video_007-030_states_LIMITED.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/MobileProjection/Phase2/014_audio_010-068_video_007-030_states_LIMITED.lua]: @JackLivio Guess the reason is the same as for previous failed script: performance degradation related to environment.

**Comment by dboltovskyi** [comment]
@JackLivio Please find scripts with reduced number of test cases in a5bdd12. It takes for about 15min. to execute them.

**Pull request merged**